File: deno/packages/tools/tools/semantic_search.ts
/**
 * Semantic Search Tool — RAG-powered codebase search.
 *
 * Provides semantic code search using vector embeddings via the
 * `@brainwires/knowledge` RAG interfaces. Supports indexing, querying,
 * filtered search, statistics, and git history search.
 *
 * Equivalent to Rust's `brainwires_tools::semantic_search` module. The Rust
 * version carries a global `OnceCell<RagClient>`; the Deno port accepts an
 * injected {@link RagClient} on each call so the host picks the transport
 * (in-process stub, HTTP RPC to the Rust service, etc.).
 */

import {
  objectSchema,
  type Tool,
  type ToolContext,
  ToolResult,
} from "@brainwires/core";
import type {
  AdvancedSearchRequest,
  IndexRequest,
  QueryRequest,
  RagClient,
  SearchGitHistoryRequest,
} from "@brainwires/knowledge";

/** Tool definitions and executor for semantic codebase search powered by RAG. */
export class SemanticSearchTool {
  /** Get all semantic search tool definitions. */
  static getTools(): Tool[] {
    return [
      {
        name: "index_codebase",
        description:
          "Index a codebase directory for semantic search using embeddings. Automatically performs full or incremental indexing. After indexing, use query_codebase to search.",
        input_schema: objectSchema({
          path: {
            type: "string",
            description: "Path to the codebase directory to index",
          },
          project: {
            type: "string",
            description:
              "Optional project name for multi-project support",
          },
          include_patterns: {
            type: "array",
            items: { type: "string" },
            description: "Optional glob patterns to include",
            default: [],
          },
          exclude_patterns: {
            type: "array",
            items: { type: "string" },
            description: "Optional glob patterns to exclude",
            default: [],
          },
          max_file_size: {
            type: "integer",
            description:
              "Maximum file size in bytes to index (default: 1MB)",
            default: 1_048_576,
          },
        }, ["path"]),
        requires_approval: false,
        defer_loading: true,
      },
      {
        name: "query_codebase",
        description:
          "Search the indexed codebase using semantic search. Returns relevant code chunks ranked by similarity.",
        input_schema: objectSchema({
          query: { type: "string", description: "The search query" },
          project: {
            type: "string",
            description: "Optional project name to filter by",
          },
          limit: {
            type: "integer",
            description: "Number of results (default: 10)",
            default: 10,
          },
          min_score: {
            type: "number",
            description:
              "Minimum similarity score 0-1 (default: 0.7)",
            default: 0.7,
          },
          hybrid: {
            type: "boolean",
            description:
              "Enable hybrid search (vector + keyword) (default: true)",
            default: true,
          },
        }, ["query"]),
        requires_approval: false,
        defer_loading: true,
      },
      {
        name: "search_with_filters",
        description:
          "Advanced semantic search with filters for file type, language, and path patterns.",
        input_schema: objectSchema({
          query: { type: "string", description: "The search query" },
          project: {
            type: "string",
            description: "Optional project name",
          },
          limit: {
            type: "integer",
            description: "Number of results (default: 10)",
            default: 10,
          },
          min_score: {
            type: "number",
            description: "Minimum score (default: 0.7)",
            default: 0.7,
          },
          file_extensions: {
            type: "array",
            items: { type: "string" },
            description: "Filter by extensions",
            default: [],
          },
          languages: {
            type: "array",
            items: { type: "string" },
            description: "Filter by languages",
            default: [],
          },
          path_patterns: {
            type: "array",
            items: { type: "string" },
            description: "Filter by path patterns",
            default: [],
          },
        }, ["query"]),
        requires_approval: false,
        defer_loading: true,
      },
      {
        name: "get_rag_statistics",
        description:
          "Get statistics about the indexed codebase (file counts, chunk counts, languages).",
        input_schema: objectSchema({
          project: {
            type: "string",
            description: "Optional project name",
          },
        }, []),
        requires_approval: false,
        defer_loading: true,
      },
      {
        name: "clear_rag_index",
        description:
          "Clear all indexed data from the vector database. Use before reindexing from scratch.",
        input_schema: objectSchema({}, []),
        requires_approval: true,
        defer_loading: true,
      },
      {
        name: "search_git_history",
        description:
          "Search git commit history using semantic search with on-demand indexing.",
        input_schema: objectSchema({
          query: { type: "string", description: "The search query" },
          path: {
            type: "string",
            description: "Path to the git repository (default: .)",
            default: ".",
          },
          project: {
            type: "string",
            description: "Optional project name",
          },
          branch: {
            type: "string",
            description: "Optional branch name",
          },
          max_commits: {
            type: "integer",
            description: "Max commits to index (default: 10)",
            default: 10,
          },
          limit: {
            type: "integer",
            description: "Number of results (default: 10)",
            default: 10,
          },
          min_score: {
            type: "number",
            description: "Minimum score (default: 0.7)",
            default: 0.7,
          },
          author: {
            type: "string",
            description: "Filter by author (regex)",
          },
          since: {
            type: "string",
            description: "Filter since date (ISO 8601)",
          },
          until: {
            type: "string",
            description: "Filter until date (ISO 8601)",
          },
          file_pattern: {
            type: "string",
            description: "Filter by file path pattern (regex)",
          },
        }, ["query"]),
        requires_approval: false,
        defer_loading: true,
      },
    ];
  }

  /** Execute a semantic search tool. */
  static async execute(
    tool_use_id: string,
    tool_name: string,
    input: Record<string, unknown>,
    _context: ToolContext,
    client: RagClient,
  ): Promise<ToolResult> {
    try {
      const output = await SemanticSearchTool.dispatch(
        tool_name,
        input,
        client,
      );
      return ToolResult.success(tool_use_id, output);
    } catch (e) {
      return ToolResult.error(
        tool_use_id,
        `Semantic search operation failed: ${(e as Error).message}`,
      );
    }
  }

  private static dispatch(
    tool_name: string,
    input: Record<string, unknown>,
    client: RagClient,
  ): Promise<string> {
    switch (tool_name) {
      case "index_codebase":
        return SemanticSearchTool.indexCodebase(input, client);
      case "query_codebase":
        return SemanticSearchTool.queryCodebase(input, client);
      case "search_with_filters":
        return SemanticSearchTool.searchWithFilters(input, client);
      case "get_rag_statistics":
        return SemanticSearchTool.getStatistics(client);
      case "clear_rag_index":
        return SemanticSearchTool.clearIndex(client);
      case "search_git_history":
        return SemanticSearchTool.searchGitHistory(input, client);
      default:
        return Promise.reject(new Error(`Unknown tool: ${tool_name}`));
    }
  }

  static async indexCodebase(
    input: Record<string, unknown>,
    client: RagClient,
  ): Promise<string> {
    const path = input.path as string | undefined;
    if (!path) throw new Error("Missing 'path' parameter");

    const req: IndexRequest = {
      path,
      project: input.project as string | undefined,
      includePatterns: Array.isArray(input.include_patterns)
        ? (input.include_patterns as string[])
        : undefined,
      excludePatterns: Array.isArray(input.exclude_patterns)
        ? (input.exclude_patterns as string[])
        : undefined,
      maxFileSize: (input.max_file_size as number | undefined) ?? 1_048_576,
    };

    const resp = await client.indexCodebase(req);
    return `Indexed ${resp.filesIndexed} files, ${resp.chunksCreated} chunks in ${resp.durationMs}ms (mode: ${resp.mode})`;
  }

  static async queryCodebase(
    input: Record<string, unknown>,
    client: RagClient,
  ): Promise<string> {
    const query = input.query as string | undefined;
    if (!query) throw new Error("Missing 'query' parameter");

    const req: QueryRequest = {
      query,
      path: input.path as string | undefined,
      project: input.project as string | undefined,
      limit: (input.limit as number | undefined) ?? 10,
      minScore: (input.min_score as number | undefined) ?? 0.7,
      hybrid: (input.hybrid as boolean | undefined) ?? true,
    };

    const resp = await client.queryCodebase(req);
    let out = `Found ${resp.results.length} results:\n\n`;
    resp.results.forEach((r, i) => {
      out += `${i + 1}. ${r.filePath} (score: ${r.score.toFixed(3)})\n`;
      out += `   Lines ${r.startLine}-${r.endLine}\n`;
      const firstLine = r.content.split("\n")[0] ?? "";
      out += `   ${firstLine}\n\n`;
    });
    return out;
  }

  static async searchWithFilters(
    input: Record<string, unknown>,
    client: RagClient,
  ): Promise<string> {
    const query = input.query as string | undefined;
    if (!query) throw new Error("Missing 'query' parameter");

    const req: AdvancedSearchRequest = {
      query,
      path: input.path as string | undefined,
      project: input.project as string | undefined,
      limit: (input.limit as number | undefined) ?? 10,
      minScore: (input.min_score as number | undefined) ?? 0.7,
      fileExtensions: Array.isArray(input.file_extensions)
        ? (input.file_extensions as string[])
        : undefined,
      languages: Array.isArray(input.languages)
        ? (input.languages as string[])
        : undefined,
      pathPatterns: Array.isArray(input.path_patterns)
        ? (input.path_patterns as string[])
        : undefined,
    };

    const resp = await client.advancedSearch(req);
    let out = `Found ${resp.results.length} filtered results:\n\n`;
    resp.results.forEach((r, i) => {
      out += `${i + 1}. ${r.filePath} (score: ${r.score.toFixed(3)})\n`;
      out += `   Language: ${r.language}\n`;
      out += `   Lines ${r.startLine}-${r.endLine}\n\n`;
    });
    return out;
  }

  static async getStatistics(client: RagClient): Promise<string> {
    const resp = await client.getStatistics();
    let out = "RAG Index Statistics:\n";
    out += `  Total chunks: ${resp.totalChunks}\n`;
    out += `  Total files: ${resp.totalFiles}\n\n`;
    if (resp.languageBreakdown.length > 0) {
      out += "Languages:\n";
      for (const s of resp.languageBreakdown) {
        out += `  ${s.language}: ${s.fileCount} files, ${s.chunkCount} chunks\n`;
      }
    }
    return out;
  }

  static async clearIndex(client: RagClient): Promise<string> {
    const resp = await client.clearIndex();
    return `Cleared index: ${resp.message}`;
  }

  static async searchGitHistory(
    input: Record<string, unknown>,
    client: RagClient,
  ): Promise<string> {
    const query = input.query as string | undefined;
    if (!query) throw new Error("Missing 'query' parameter");

    const req: SearchGitHistoryRequest = {
      query,
      path: (input.path as string | undefined) ?? ".",
      project: input.project as string | undefined,
      branch: input.branch as string | undefined,
      maxCommits: (input.max_commits as number | undefined) ?? 10,
      limit: (input.limit as number | undefined) ?? 10,
      minScore: (input.min_score as number | undefined) ?? 0.7,
      author: input.author as string | undefined,
      since: input.since as string | undefined,
      until: input.until as string | undefined,
      filePattern: input.file_pattern as string | undefined,
    };

    const resp = await client.searchGitHistory(req);
    let out = `Found ${resp.results.length} commits:\n\n`;
    resp.results.forEach((r, i) => {
      out += `${i + 1}. ${r.commitHash.slice(0, 8)} (score: ${
        r.score.toFixed(3)
      })\n`;
      out += `   Author: ${r.author}\n`;
      out += `   Date: ${r.commitDate}\n`;
      out += `   Message: ${r.commitMessage}\n\n`;
    });
    return out;
  }
}

File: crates/brainwires-tool-builtins/src/semantic_search.rs
//! Semantic Search Tool - RAG-powered codebase search
//!
//! Provides semantic code search using vector embeddings via the brainwires-rag crate.
//! Supports indexing, querying, filtered search, statistics, and git history search.
//!
//! Requires the `rag` feature flag.

use anyhow::Result;
use brainwires_rag::{
    AdvancedSearchRequest, IndexRequest, QueryRequest, RagClient, SearchGitHistoryRequest,
};
use serde_json::{Value, json};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::OnceCell;

use brainwires_core::{Tool, ToolContext, ToolInputSchema, ToolResult};

/// Global RAG client instance (lazy initialized)
static RAG_CLIENT: OnceCell<Arc<RagClient>> = OnceCell::const_new();

/// Get or initialize the RAG client
async fn get_rag_client() -> Result<Arc<RagClient>> {
    RAG_CLIENT
        .get_or_try_init(|| async {
            let client = RagClient::new().await?;
            Ok(Arc::new(client))
        })
        .await
        .map(Arc::clone)
}

/// Tool definitions and executor for semantic codebase search powered by RAG.
pub struct SemanticSearchTool;

impl SemanticSearchTool {
    /// Get all semantic search tool definitions
    pub fn get_tools() -> Vec<Tool> {
        let mut index_properties = HashMap::new();
        index_properties.insert(
            "path".to_string(),
            json!({"type": "string", "description": "Path to the codebase directory to index"}),
        );
        index_properties.insert(
            "project".to_string(),
            json!({"type": "string", "description": "Optional project name for multi-project support"}),
        );
        index_properties.insert(
            "include_patterns".to_string(),
            json!({"type": "array", "items": {"type": "string"}, "description": "Optional glob patterns to include", "default": []}),
        );
        index_properties.insert(
            "exclude_patterns".to_string(),
            json!({"type": "array", "items": {"type": "string"}, "description": "Optional glob patterns to exclude", "default": []}),
        );
        index_properties.insert(
            "max_file_size".to_string(),
            json!({"type": "integer", "description": "Maximum file size in bytes to index (default: 1MB)", "default": 1048576}),
        );

        vec![
            Tool {
                name: "index_codebase".to_string(),
                description: "Index a codebase directory for semantic search using embeddings. Automatically performs full or incremental indexing. After indexing, use query_codebase to search.".to_string(),
                input_schema: ToolInputSchema::object(index_properties, vec!["path".to_string()]),
                requires_approval: false,
                defer_loading: true,
                ..Default::default()
            },
            Tool {
                name: "query_codebase".to_string(),
                description: "Search the indexed codebase using semantic search. Returns relevant code chunks ranked by similarity.".to_string(),
                input_schema: {
                    let mut properties = HashMap::new();
                    properties.insert("query".to_string(), json!({"type": "string", "description": "The search query"}));
                    properties.insert("project".to_string(), json!({"type": "string", "description": "Optional project name to filter by"}));
                    properties.insert("limit".to_string(), json!({"type": "integer", "description": "Number of results (default: 10)", "default": 10}));
                    properties.insert("min_score".to_string(), json!({"type": "number", "description": "Minimum similarity score 0-1 (default: 0.7)", "default": 0.7}));
                    properties.insert("hybrid".to_string(), json!({"type": "boolean", "description": "Enable hybrid search (vector + keyword) (default: true)", "default": true}));
                    ToolInputSchema::object(properties, vec!["query".to_string()])
                },
                requires_approval: false,
                defer_loading: true,
                ..Default::default()
            },
            Tool {
                name: "search_with_filters".to_string(),
                description: "Advanced semantic search with filters for file type, language, and path patterns.".to_string(),
                input_schema: {
                    let mut properties = HashMap::new();
                    properties.insert("query".to_string(), json!({"type": "string", "description": "The search query"}));
                    properties.insert("project".to_string(), json!({"type": "string", "description": "Optional project name"}));
                    properties.insert("limit".to_string(), json!({"type": "integer", "description": "Number of results (default: 10)", "default": 10}));
                    properties.insert("min_score".to_string(), json!({"type": "number", "description": "Minimum score (default: 0.7)", "default": 0.7}));
                    properties.insert("file_extensions".to_string(), json!({"type": "array", "items": {"type": "string"}, "description": "Filter by extensions", "default": []}));
                    properties.insert("languages".to_string(), json!({"type": "array", "items": {"type": "string"}, "description": "Filter by languages", "default": []}));
                    properties.insert("path_patterns".to_string(), json!({"type": "array", "items": {"type": "string"}, "description": "Filter by path patterns", "default": []}));
                    ToolInputSchema::object(properties, vec!["query".to_string()])
                },
                requires_approval: false,
                defer_loading: true,
                ..Default::default()
            },
            Tool {
                name: "get_rag_statistics".to_string(),
                description: "Get statistics about the indexed codebase (file counts, chunk counts, languages).".to_string(),
                input_schema: {
                    let mut properties = HashMap::new();
                    properties.insert("project".to_string(), json!({"type": "string", "description": "Optional project name"}));
                    ToolInputSchema::object(properties, vec![])
                },
                requires_approval: false,
                defer_loading: true,
                ..Default::default()
            },
            Tool {
                name: "clear_rag_index".to_string(),
                description: "Clear all indexed data from the vector database. Use before reindexing from scratch.".to_string(),
                input_schema: ToolInputSchema::object(HashMap::new(), vec![]),
                requires_approval: true,
                defer_loading: true,
                ..Default::default()
            },
            Tool {
                name: "search_git_history".to_string(),
                description: "Search git commit history using semantic search with on-demand indexing.".to_string(),
                input_schema: {
                    let mut properties = HashMap::new();
                    properties.insert("query".to_string(), json!({"type": "string", "description": "The search query"}));
                    properties.insert("path".to_string(), json!({"type": "string", "description": "Path to the git repository (default: .)", "default": "."}));
                    properties.insert("project".to_string(), json!({"type": "string", "description": "Optional project name"}));
                    properties.insert("branch".to_string(), json!({"type": "string", "description": "Optional branch name"}));
                    properties.insert("max_commits".to_string(), json!({"type": "integer", "description": "Max commits to index (default: 10)", "default": 10}));
                    properties.insert("limit".to_string(), json!({"type": "integer", "description": "Number of results (default: 10)", "default": 10}));
                    properties.insert("min_score".to_string(), json!({"type": "number", "description": "Minimum score (default: 0.7)", "default": 0.7}));
                    properties.insert("author".to_string(), json!({"type": "string", "description": "Filter by author (regex)"}));
                    properties.insert("since".to_string(), json!({"type": "string", "description": "Filter since date (ISO 8601)"}));
                    properties.insert("until".to_string(), json!({"type": "string", "description": "Filter until date (ISO 8601)"}));
                    properties.insert("file_pattern".to_string(), json!({"type": "string", "description": "Filter by file path pattern (regex)"}));
                    ToolInputSchema::object(properties, vec!["query".to_string()])
                },
                requires_approval: false,
                defer_loading: true,
                ..Default::default()
            },
        ]
    }

    /// Execute a semantic search tool
    pub async fn execute(
        tool_use_id: &str,
        tool_name: &str,
        input: &Value,
        _context: &ToolContext,
    ) -> ToolResult {
        let result = match tool_name {
            "index_codebase" => Self::index_codebase(input).await,
            "query_codebase" => Self::query_codebase(input).await,
            "search_with_filters" => Self::search_with_filters(input).await,
            "get_rag_statistics" => Self::get_statistics(input).await,
            "clear_rag_index" => Self::clear_index(input).await,
            "search_git_history" => Self::search_git_history(input).await,
            _ => Err(anyhow::anyhow!("Unknown tool: {}", tool_name)),
        };

        match result {
            Ok(output) => ToolResult::success(tool_use_id.to_string(), output),
            Err(e) => ToolResult::error(
                tool_use_id.to_string(),
                format!("Semantic search operation failed: {}", e),
            ),
        }
    }

    async fn index_codebase(input: &Value) -> Result<String> {
        let client = get_rag_client().await?;

        let path = input["path"]
            .as_str()
            .ok_or_else(|| anyhow::anyhow!("Missing 'path' parameter"))?
            .to_string();

        let request = IndexRequest {
            path,
            project: input["project"].as_str().map(|s| s.to_string()),
            include_patterns: input["include_patterns"]
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(|s| s.to_string()))
                        .collect()
                })
                .unwrap_or_default(),
            exclude_patterns: input["exclude_patterns"]
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(|s| s.to_string()))
                        .collect()
                })
                .unwrap_or_default(),
            max_file_size: input["max_file_size"].as_u64().unwrap_or(1_048_576) as usize,
        };

        let response = client.index_codebase(request).await?;

        Ok(format!(
            "Indexed {} files, {} chunks in {}ms (mode: {:?})",
            response.files_indexed, response.chunks_created, response.duration_ms, response.mode
        ))
    }

    async fn query_codebase(input: &Value) -> Result<String> {
        let client = get_rag_client().await?;

        let request = QueryRequest {
            query: input["query"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("Missing 'query' parameter"))?
                .to_string(),
            path: input["path"].as_str().map(|s| s.to_string()),
            project: input["project"].as_str().map(|s| s.to_string()),
            limit: input["limit"].as_u64().unwrap_or(10) as usize,
            min_score: input["min_score"].as_f64().unwrap_or(0.7) as f32,
            hybrid: input["hybrid"].as_bool().unwrap_or(true),
        };

        let response = client.query_codebase(request).await?;

        let mut output = format!("Found {} results:\n\n", response.results.len());
        for (i, result) in response.results.iter().enumerate() {
            output.push_str(&format!(
                "{}. {} (score: {:.3})\n",
                i + 1,
                result.file_path,
                result.score
            ));
            output.push_str(&format!(
                "   Lines {}-{}\n",
                result.start_line, result.end_line
            ));
            output.push_str(&format!(
                "   {}\n\n",
                result.content.lines().next().unwrap_or("")
            ));
        }

        Ok(output)
    }

    async fn search_with_filters(input: &Value) -> Result<String> {
        let client = get_rag_client().await?;

        let request = AdvancedSearchRequest {
            query: input["query"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("Missing 'query' parameter"))?
                .to_string(),
            path: input["path"].as_str().map(|s| s.to_string()),
            project: input["project"].as_str().map(|s| s.to_string()),
            limit: input["limit"].as_u64().unwrap_or(10) as usize,
            min_score: input["min_score"].as_f64().unwrap_or(0.7) as f32,
            file_extensions: input["file_extensions"]
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(|s| s.to_string()))
                        .collect()
                })
                .unwrap_or_default(),
            languages: input["languages"]
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(|s| s.to_string()))
                        .collect()
                })
                .unwrap_or_default(),
            path_patterns: input["path_patterns"]
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(|s| s.to_string()))
                        .collect()
                })
                .unwrap_or_default(),
        };

        let response = client.search_with_filters(request).await?;

        let mut output = format!("Found {} filtered results:\n\n", response.results.len());
        for (i, result) in response.results.iter().enumerate() {
            output.push_str(&format!(
                "{}. {} (score: {:.3})\n",
                i + 1,
                result.file_path,
                result.score
            ));
            output.push_str(&format!("   Language: {}\n", result.language));
            output.push_str(&format!(
                "   Lines {}-{}\n\n",
                result.start_line, result.end_line
            ));
        }

        Ok(output)
    }

    async fn get_statistics(_input: &Value) -> Result<String> {
        let client = get_rag_client().await?;

        let response = client.get_statistics().await?;

        let mut output = String::from("RAG Index Statistics:\n");
        output.push_str(&format!("  Total chunks: {}\n", response.total_chunks));
        output.push_str(&format!("  Total files: {}\n\n", response.total_files));

        if !response.language_breakdown.is_empty() {
            output.push_str("Languages:\n");
            for lang_stat in &response.language_breakdown {
                output.push_str(&format!(
                    "  {}: {} files, {} chunks\n",
                    lang_stat.language, lang_stat.file_count, lang_stat.chunk_count
                ));
            }
        }

        Ok(output)
    }

    async fn clear_index(_input: &Value) -> Result<String> {
        let client = get_rag_client().await?;

        let response = client.clear_index().await?;

        Ok(format!("Cleared index: {}", response.message))
    }

    async fn search_git_history(input: &Value) -> Result<String> {
        let client = get_rag_client().await?;

        let request = SearchGitHistoryRequest {
            query: input["query"]
                .as_str()
                .ok_or_else(|| anyhow::anyhow!("Missing 'query' parameter"))?
                .to_string(),
            path: input["path"].as_str().unwrap_or(".").to_string(),
            project: input["project"].as_str().map(|s| s.to_string()),
            branch: input["branch"].as_str().map(|s| s.to_string()),
            max_commits: input["max_commits"].as_u64().unwrap_or(10) as usize,
            limit: input["limit"].as_u64().unwrap_or(10) as usize,
            min_score: input["min_score"].as_f64().unwrap_or(0.7) as f32,
            author: input["author"].as_str().map(|s| s.to_string()),
            since: input["since"].as_str().map(|s| s.to_string()),
            until: input["until"].as_str().map(|s| s.to_string()),
            file_pattern: input["file_pattern"].as_str().map(|s| s.to_string()),
        };

        let response = client.search_git_history(request).await?;

        let mut output = format!("Found {} commits:\n\n", response.results.len());
        for (i, result) in response.results.iter().enumerate() {
            output.push_str(&format!(
                "{}. {} (score: {:.3})\n",
                i + 1,
                &result.commit_hash[..8],
                result.score
            ));
            output.push_str(&format!("   Author: {}\n", result.author));
            output.push_str(&format!("   Date: {}\n", result.commit_date));
            output.push_str(&format!("   Message: {}\n\n", result.commit_message));
        }

        Ok(output)
    }

    // ============ Helper methods for orchestrator integration ============

    /// Execute a query against the indexed codebase (for orchestrator)
    pub async fn execute_query(
        query: &str,
        limit: usize,
        min_score: f32,
    ) -> Result<String, String> {
        let input = json!({
            "query": query,
            "limit": limit,
            "min_score": min_score,
            "hybrid": true
        });
        Self::query_codebase(&input)
            .await
            .map_err(|e| e.to_string())
    }

    /// Index a codebase directory (for orchestrator)
    pub async fn execute_index(path: &str) -> Result<String, String> {
        let input = json!({ "path": path });
        Self::index_codebase(&input)
            .await
            .map_err(|e| e.to_string())
    }

    /// Execute filtered search (for orchestrator)
    pub async fn execute_filtered_search(input: &Value) -> Result<String, String> {
        Self::search_with_filters(input)
            .await
            .map_err(|e| e.to_string())
    }

    /// Get RAG statistics (for orchestrator)
    pub async fn execute_get_stats() -> Result<String, String> {
        Self::get_statistics(&Value::Null)
            .await
            .map_err(|e| e.to_string())
    }

    /// Search git history (for orchestrator)
    pub async fn execute_git_history_search(input: &Value) -> Result<String, String> {
        Self::search_git_history(input)
            .await
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_tools() {
        let tools = SemanticSearchTool::get_tools();
        assert_eq!(tools.len(), 6);

        let tool_names: Vec<&str> = tools.iter().map(|t| t.name.as_str()).collect();
        assert!(tool_names.contains(&"index_codebase"));
        assert!(tool_names.contains(&"query_codebase"));
        assert!(tool_names.contains(&"search_with_filters"));
        assert!(tool_names.contains(&"get_rag_statistics"));
        assert!(tool_names.contains(&"clear_rag_index"));
        assert!(tool_names.contains(&"search_git_history"));
    }

    #[test]
    fn test_index_codebase_tool_definition() {
        let tools = SemanticSearchTool::get_tools();
        let index_tool = tools.iter().find(|t| t.name == "index_codebase").unwrap();

        assert!(index_tool.description.contains("Index"));
        assert!(index_tool.description.contains("semantic"));
        assert!(!index_tool.requires_approval);
        assert!(index_tool.defer_loading);
    }

    #[test]
    fn test_query_codebase_tool_definition() {
        let tools = SemanticSearchTool::get_tools();
        let query_tool = tools.iter().find(|t| t.name == "query_codebase").unwrap();

        assert!(query_tool.description.contains("Search"));
        assert!(!query_tool.requires_approval);
    }

    #[test]
    fn test_clear_index_requires_approval() {
        let tools = SemanticSearchTool::get_tools();
        let clear_tool = tools.iter().find(|t| t.name == "clear_rag_index").unwrap();
        assert!(clear_tool.requires_approval);
    }

    #[test]
    fn test_all_tools_have_descriptions() {
        let tools = SemanticSearchTool::get_tools();
        for tool in tools {
            assert!(
                !tool.description.is_empty(),
                "Tool {} should have a description",
                tool.name
            );
        }
    }

    #[test]
    fn test_index_codebase_required_fields() {
        let tools = SemanticSearchTool::get_tools();
        let index_tool = tools.iter().find(|t| t.name == "index_codebase").unwrap();

        if let Some(ref required) = index_tool.input_schema.required {
            assert!(required.contains(&"path".to_string()));
        }
    }
}
